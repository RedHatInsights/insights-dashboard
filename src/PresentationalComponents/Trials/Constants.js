import React from 'react';
import trialsMessages from './TrialsMessages';
import TrialsText from './TrialsText';
import TrialsBullets from './TrialsBullets';
import { Button, TextContent } from '@patternfly/react-core';
import NoConnectedSystemsTable from './NoConnectedSystemsTable';
import NoEligibleSystemsTable from './NoEligibleSystemsTable';

import { intl } from '../../Utilities/IntlProvider';

const rhcConnectedFilter = [
    {
        rhcdFilter: ['not_nil']
    }
];

const rhcEligibleFilter = [
    {
        rhcdFilter: ['nil']
    },
    {
        osFilter: ['8.6', '8.7', '8.8', '9.0', '9.1', '9.2']
    }
];

const learnMoreMessage = {
    message: trialsMessages.systemsEligibleDescriptionTwo,
    value: {
        a: (chunks) => <a href={`https://access.redhat.com/articles/rhc-registration`}>{ chunks }</a>
    }
};

const systemsEligibleDescription =
    <TextContent><TrialsText messages={[trialsMessages.systemsEligibleDescriptionOne, learnMoreMessage]} /></TextContent>;

const setSystemTable = (
    setIsModalOpened,
    setDefaultActiveFilters,
    setActiveEmptyState,
    emptyState,
    filter,
    setModalTitle,
    title,
    setModalDescription,
    description
) => {
    setIsModalOpened(true);
    setDefaultActiveFilters(filter);
    setActiveEmptyState(emptyState);
    setModalTitle(title);
    setModalDescription(description);
};

export const faqs = [
    {
        faq: trialsMessages.faqOne,
        answer: <TextContent><TrialsText messages={[trialsMessages.faqOneAnswer]} style={{ color: '#6a6e73' }} /></TextContent>
    },
    {
        faq: trialsMessages.faqTwo,
        answer: <TextContent><TrialsText messages={[trialsMessages.faqTwoAnswer]} style={{ color: '#6a6e73' }} /></TextContent>
    },
    {
        faq: trialsMessages.faqThree,
        answer:
            <TextContent>
                <TrialsText
                    messages={[trialsMessages.faqThreeAnswerParagraphOne, trialsMessages.faqThreeAnswerParagraphTwo]}
                    style={{ color: '#6a6e73' }}
                />
                <TrialsBullets
                    isNumbered={false}
                    messages={[
                        intl.formatMessage(trialsMessages.faqThreeAnswerBulletOne),
                        intl.formatMessage(trialsMessages.faqThreeAnswerBulletTwo),
                        intl.formatMessage(trialsMessages.faqThreeAnswerBulletThree)
                    ]}
                    style={{ color: '#6a6e73' }}
                />
                <TrialsText messages={[trialsMessages.faqThreeAnswerParagraphThree]} color='#6a6e73' />
            </TextContent>
    },
    {
        faq: trialsMessages.faqFour,
        answer: <TextContent><TrialsText messages={[trialsMessages.faqFourAnswer]} style={{ color: '#6a6e73' }} /></TextContent>
    },
    {
        faq: trialsMessages.faqFive,
        answer: <TextContent><TrialsText messages={[trialsMessages.faqFiveAnswer]} style={{ color: '#6a6e73' }} /></TextContent>
    },
    {
        faq: trialsMessages.faqSix,
        answer: <TextContent><TrialsText messages={[trialsMessages.faqSixAnswer]} style={{ color: '#6a6e73' }} /></TextContent>
    },
    {
        faq: trialsMessages.faqSeven,
        answer:
            <TextContent>
                <TrialsText messages={[trialsMessages.faqSevenAnswerParagraphOne]} style={{ color: '#6a6e73' }} />
                <TrialsBullets
                    isNumbered={true}
                    messages={[
                        intl.formatMessage(trialsMessages.faqSevenAnswerBulletOne),
                        intl.formatMessage(trialsMessages.faqSevenAnswerBulletTwo),
                        intl.formatMessage(trialsMessages.faqSevenAnswerBulletThree),
                        intl.formatMessage(trialsMessages.faqSevenAnswerBulletFour)
                    ]}
                    style={{ color: '#6a6e73' }}
                />
            </TextContent>
    },
    {
        faq: trialsMessages.faqEight,
        answer: <TextContent><TrialsText messages={[trialsMessages.faqEightAnswer]} style={{ color: '#6a6e73' }} /></TextContent>
    },
    {
        faq: trialsMessages.faqNine,
        answer:
            <TextContent>
                <TrialsBullets
                    isNumbered={false}
                    messages={[
                        intl.formatMessage(trialsMessages.faqNineAnswerBulletOne),
                        intl.formatMessage(trialsMessages.faqNineAnswerBulletTwo),
                        intl.formatMessage(trialsMessages.faqNineAnswerBulletThree)
                    ]}
                    style={{ color: '#6a6e73' }}
                />
            </TextContent>
    }
];

export const gettingStartedBullets = (setIsModalOpened, setDefaultActiveFilters, setActiveEmptyState, setModalTitle, setModalDescription) => {
    return [
        {
            title: trialsMessages.gettingStartedBulletOneTitle,
            message: <TextContent>
                <TrialsBullets
                    isNumbered={false}
                    messages={[
                        intl.formatMessage(trialsMessages.discoverSystemsBulletOne),
                        <Button
                            key='rhc-connected-system-link'
                            style={{ padding: '0px' }}
                            variant="link"
                            onClick={() => setSystemTable(
                                setIsModalOpened,
                                setDefaultActiveFilters,
                                setActiveEmptyState,
                                NoConnectedSystemsTable,
                                rhcConnectedFilter,
                                setModalTitle,
                                intl.formatMessage(trialsMessages.systemsConnectedTitle),
                                setModalDescription
                            )}
                        >
                            {intl.formatMessage(trialsMessages.discoverSystemsBulletTwo)}
                        </Button>,
                        <Button
                            key='rhc-eligible-system-link'
                            style={{ padding: '0px' }}
                            variant="link"
                            onClick={() => setSystemTable(
                                setIsModalOpened,
                                setDefaultActiveFilters,
                                setActiveEmptyState,
                                NoEligibleSystemsTable,
                                rhcEligibleFilter,
                                setModalTitle,
                                intl.formatMessage(trialsMessages.systemsEligibleTitle),
                                setModalDescription,
                                systemsEligibleDescription
                            )}
                        >
                            {intl.formatMessage(trialsMessages.discoverSystemsBulletThree)}
                        </Button>,
                        <React.Fragment key="getting-started-discover-systems">
                            {intl.formatMessage(trialsMessages.discoverSystemsBulletFour)}
                            <TrialsBullets
                                isNumbered={false}
                                messages={[
                                    intl.formatMessage(trialsMessages.connectRHCSystemsBulletOne, {
                                        a: (chunks) => <a href={`https://access.redhat.com/articles/rhc-registration`}>{ chunks }</a>
                                    })
                                ]}
                            />
                        </React.Fragment>
                    ]}
                />
            </TextContent>,
            number: 1
        },
        {
            title: trialsMessages.gettingStartedBulletTwoTitle,
            message: <TextContent>
                <TrialsBullets
                    isNumbered={false}
                    messages={[
                        intl.formatMessage(trialsMessages.updateRHCSettingsBulletOne)
                    ]}
                />
            </TextContent>,
            number: 2
        },
        {
            title: trialsMessages.gettingStartedBulletThreeTitle,
            message: <TextContent>
                <TrialsBullets
                    isNumbered={false}
                    messages={[
                        intl.formatMessage(trialsMessages.createPlaybookBulletOne),
                        intl.formatMessage(trialsMessages.createPlaybookBulletTwo),
                        intl.formatMessage(trialsMessages.createPlaybookBulletThree)
                    ]}
                />
            </TextContent>,
            number: 3
        }
    ];
};
