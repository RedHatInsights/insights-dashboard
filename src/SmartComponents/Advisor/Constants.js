export const INCIDENT_URL = `/advisor/recommendations?reports_shown=undefined&impacting=false&offset=0&limit=10&sort=-publish_date&incident=true`;
export const NEW_REC_URL = `/advisor/recommendations?reports_shown=undefined&impacting=false&offset=0&limit=10&sort=-publish_date`;

export const SEVERITY_MAP = {
    critical: 4,
    important: 3,
    moderate: 2,
    low: 1
};
