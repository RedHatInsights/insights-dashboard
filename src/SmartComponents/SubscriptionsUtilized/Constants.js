export const RHSM_API_RESPONSE_DATA = 'data';

export const RHSM_API_RESPONSE_DATA_TYPES = {
    CORES: 'cores',
    DATE: 'date',
    SOCKETS: 'sockets',
    HAS_DATA: 'has_data',
    HAS_INFINITE: 'has_infinite_quantity'
};

export const RHSM_API_PRODUCT_ID_TYPES = {
    RHEL: 'RHEL',
    OPENSHIFT: 'OpenShift Container Platform'
};

export const RHSM_API_QUERY_GRANULARITY_TYPES = {
    DAILY: 'daily'
};
