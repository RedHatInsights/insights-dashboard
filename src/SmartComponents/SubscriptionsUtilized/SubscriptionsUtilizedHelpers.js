import moment from 'moment';
import { RHSM_API_RESPONSE_DATA_TYPES } from './Constants';

/**
 * Generate a range of dates.
 *
 * @param {Date} date
 * @param {number} subtract
 * @param {string} measurement
 * @returns {{endDate: Date, startDate: Date}}
 */
export const setRangedDateTime = (date = new Date(), subtract = 1, measurement = 'days') => ({
    startDate: moment
    .utc(date)
    .startOf(measurement)
    .subtract(subtract, measurement)
    .toDate(),
    endDate: moment
    .utc(date)
    .startOf(measurement)
    .endOf('days')
    .toDate()
});

/**
 * Apply a set of schemas using either an array of objects in the
 * form of [{ madeUpKey: 'some_api_key' }], or an array of arrays
 * in the form of [['some_api_key','another_api_key']]
 *
 * @param {Array} schemas
 * @param {*} initialValue
 * @returns {Array}
 */
export const setResponseSchemas = (schemas = [], initialValue) =>
    schemas.map(schema => {
        const generated = {};
        const arr = (Array.isArray(schema) && schema) || Object.values(schema);
        arr.forEach(value => {
            generated[value] = initialValue;
        });
        return generated;
    });

/**
 * Filter report and capacity data against expected API response. Apply percentage and date.
 *
 * @param {Array} report
 * @param {Array} capacity
 * @param {string} filter
 * @return {({}|{date: Date, report: (number|null|undefined), capacity: (number|null|undefined),
 *     percentage: (number|null|undefined)})}
 */
export const filterChartData = (report = [], capacity = [], filter) => {
    const reportData = report.reverse();
    const capacityData = capacity.reverse();
    let chartData = {};
    let recordIndex = null;

    const [responseSchema = {}] = setResponseSchemas([
        RHSM_API_RESPONSE_DATA_TYPES
    ]);

    const record = reportData.find((data, index) => {
        recordIndex = index;
        return data[RHSM_API_RESPONSE_DATA_TYPES.HAS_DATA] === true;
    });

    if (record) {
        const date = record[RHSM_API_RESPONSE_DATA_TYPES.DATE];
        chartData = {
            date,
            report: { ...responseSchema, ...record },
            capacity: { ...responseSchema, ...capacityData[recordIndex] },
            percentage: undefined
        };

        chartData.report = chartData.report[filter];
        chartData.capacity = chartData.capacity[RHSM_API_RESPONSE_DATA_TYPES.HAS_INFINITE] ? null : chartData.capacity[filter];
        let percentage = chartData.capacity === null ? null : ((chartData.report || 0) / (chartData.capacity || 0)) * 100;

        if (Number.isNaN(percentage)) {
            percentage = 0;
        }

        if (!Number.isFinite(percentage)) {
            percentage = undefined;
        }

        if (typeof percentage === 'number') {
            percentage = Math.ceil(percentage);
        }

        chartData.percentage = percentage;
    }

    return chartData;
};
