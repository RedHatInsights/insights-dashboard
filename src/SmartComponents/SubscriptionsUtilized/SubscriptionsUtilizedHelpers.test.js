import { setRangedDateTime, setResponseSchemas, filterChartData } from './SubscriptionsUtilizedHelpers';
import * as RHSM_TYPES from './Constants';

describe('SmartComponents/SubscriptionsUtilized/SubscriptionsUtilizedHelpers', () => {
    it('should return a range of date and time that includes the prior date', () => {
        expect(setRangedDateTime('20200401')).toMatchSnapshot('date and time');
    });

    it('should return a predictable schema object', () => {
        expect(setResponseSchemas([
            RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES
        ])).toMatchSnapshot('schema');
    });

    it('should return filtered chart data', () => {
        const reportResponse = {
            [RHSM_TYPES.RHSM_API_RESPONSE_DATA]: [{
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.SOCKETS]: 300,
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.CORES]: 100,
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.HAS_DATA]: true
            }, {
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.SOCKETS]: 0,
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.CORES]: 0,
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.HAS_DATA]: false
            }]
        };
        const capacityResponse = {
            [RHSM_TYPES.RHSM_API_RESPONSE_DATA]: [{
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.SOCKETS]: 10,
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.CORES]: 0,
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.HAS_INFINITE]: false
            }, {
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.SOCKETS]: 0,
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.CORES]: 0,
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.HAS_INFINITE]: false
            }]
        };

        expect(filterChartData(reportResponse[RHSM_TYPES.RHSM_API_RESPONSE_DATA],
            capacityResponse[RHSM_TYPES.RHSM_API_RESPONSE_DATA],
            [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.SOCKETS])).toMatchSnapshot('filtered');
    });

    it('should return null and undefined percentages based on specific capacity values', () => {
        const reportResponse = {
            [RHSM_TYPES.RHSM_API_RESPONSE_DATA]: [{
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.SOCKETS]: 300,
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.HAS_DATA]: true
            }]
        };
        const capacityResponse = {
            [RHSM_TYPES.RHSM_API_RESPONSE_DATA]: [{
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.SOCKETS]: 0,
                [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.HAS_INFINITE]: false
            }]
        };

        expect(filterChartData(reportResponse[RHSM_TYPES.RHSM_API_RESPONSE_DATA],
            capacityResponse[RHSM_TYPES.RHSM_API_RESPONSE_DATA],
            [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.SOCKETS]).percentage).toBe(undefined);

        capacityResponse[RHSM_TYPES.RHSM_API_RESPONSE_DATA][0][RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.SOCKETS] = null;

        expect(filterChartData(reportResponse[RHSM_TYPES.RHSM_API_RESPONSE_DATA],
            capacityResponse[RHSM_TYPES.RHSM_API_RESPONSE_DATA],
            [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.SOCKETS]).percentage).toBe(null);

        capacityResponse[RHSM_TYPES.RHSM_API_RESPONSE_DATA][0][RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.SOCKETS] = 0;
        capacityResponse[RHSM_TYPES.RHSM_API_RESPONSE_DATA][0][RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.HAS_INFINITE] = true;

        expect(filterChartData(reportResponse[RHSM_TYPES.RHSM_API_RESPONSE_DATA],
            capacityResponse[RHSM_TYPES.RHSM_API_RESPONSE_DATA],
            [RHSM_TYPES.RHSM_API_RESPONSE_DATA_TYPES.SOCKETS]).percentage).toBe(null);
    });
});
