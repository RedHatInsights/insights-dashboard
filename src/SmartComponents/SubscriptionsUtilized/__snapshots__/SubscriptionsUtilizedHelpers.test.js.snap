// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmartComponents/SubscriptionsUtilized/SubscriptionsUtilizedHelpers should return a predictable schema object: schema 1`] = `
Array [
  Object {
    "cores": undefined,
    "date": undefined,
    "has_data": undefined,
    "has_infinite_quantity": undefined,
    "sockets": undefined,
  },
]
`;

exports[`SmartComponents/SubscriptionsUtilized/SubscriptionsUtilizedHelpers should return a range of date and time that includes the prior date: date and time 1`] = `
Object {
  "endDate": 2020-04-01T23:59:59.999Z,
  "startDate": 2020-03-31T00:00:00.000Z,
}
`;

exports[`SmartComponents/SubscriptionsUtilized/SubscriptionsUtilizedHelpers should return filtered chart data: filtered 1`] = `
Object {
  "capacity": 10,
  "date": undefined,
  "percentage": 3000,
  "report": 300,
}
`;

exports[`SmartComponents/SubscriptionsUtilized/SubscriptionsUtilizedHelpers should return zeroed data when HAS_DATA is true: hasData FALSE 1`] = `Object {}`;

exports[`SmartComponents/SubscriptionsUtilized/SubscriptionsUtilizedHelpers should return zeroed data when HAS_DATA is true: hasData TRUE 1`] = `
Object {
  "capacity": 0,
  "date": undefined,
  "percentage": 0,
  "report": 0,
}
`;
